pkg/controller/configmap/configmap.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2020 Hewlett Packard Enterprise Development LP
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configmap
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	kdv1 "github.com/bluek8s/kubedirector/pkg/apis/kubedirector/v1beta1"
+	"github.com/bluek8s/kubedirector/pkg/observer"
+	"github.com/bluek8s/kubedirector/pkg/shared"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// ConfigMapType is a label placed on desired comfig maps that
+	// we want to watch and propogate inside containers
+	configMapType = shared.KdDomainBase + "/cmType"
+)
+
+// syncConfigMap runs the reconciliation logic. It is invoked because of a
+// change in or addition of configmap instance, currently there is no
+// polling for this resource. If the configmap is not labeled
+// with key ConfigMapType then no op
+func (r *ReconcileConfigMap) syncConfigMap(
+	reqLogger logr.Logger,
+	configmap *corev1.ConfigMap,
+) error {
+
+	// Memoize state of the incoming object.
+	oldMap, _ := observer.GetConfigMap(configmap.Namespace, configmap.Name)
+	if _, ok := oldMap.Labels[configMapType]; !ok {
+		return nil
+	}
+	/* anonymous fun to check if some cluster
+	   is using this config map as a connection */
+	isClusterUsingConfigMap := func(cmName string, cluster kdv1.KubeDirectorCluster) bool {
+		clusterModels := cluster.Spec.Connections.ConfigMaps
+		for _, modelMapName := range clusterModels {
+			if modelMapName == cmName {
+				return true
+			}
+		}
+		return false
+	}
+	allClusters := &kdv1.KubeDirectorClusterList{}
+	shared.List(context.TODO(), allClusters)
+	for _, kubecluster := range allClusters.Items {
+		if isClusterUsingConfigMap(configmap.Name, kubecluster) {
+			shared.LogInfof(
+				reqLogger,
+				configmap,
+				shared.EventReasonConfigMap,
+				"configmap {%s} is a connection to cluster {%s}, updating its configmeta",
+				configmap.Name,
+				kubecluster.Name,
+			)
+
+			//Notify cluster by incrementing configmetaGenerator
+			wait := time.Second
+			maxWait := 4096 * time.Second
+			for {
+				// Get a fresh copy of this cluster to work with and
+				// try update
```

Comment:
The first time through this loop (which will probably succeed) we do already have a copy we just loaded. We only need to re-load and re-apply our spec inc if the Update fails. That will cut the number of GETs by 1/2 in the common case.

Same comment for the other similar loop.

Reply:
I guess to be clear, what I'm saying is:

- coming into this code here we already have kubecluster
- increment its spec gen number
- try the shared.Update
- if that succeeds, exit the loop
- otherwise do the shared.Get to load a fresh copy into kubecluster and loop back around

Reply:
Another thing to be aware of is that the Get could fail. If it fails because the cluster is gone, no biggy, just exit the loop. If it fails for other reasons I guess we should keep looping until maxWait is exceeded.

FWIW the status-update loop in the defer func in syncCluster has to deal with basically this same sort of flow.

Reply:
Yes, already took care of above 2, there is a race between changes and reviewing :)

Reply:
:+1:

I see that in the GET error case (other than not-found) you are returning the error rather than looping to try again. Thinking about that at the moment. I guess the question is how persistent we want to be vs. willing to drop an update. (Not that a GET error is very likely.)

Reply:
👍 👍 👍